src/builtins.cpp

```diff
@@ -321,957 +235,346 @@ static void lUpdateIntrinsicsAttributes(llvm::Module *module) {
-    mapped to ispc functions are also added to the symbol table.
-
-    @param lib         Pointer to BitcodeLib class representing LLVM bitcode (e.g. the contents of a *.bc file)
-    @param module      Module to link the bitcode into
-    @param symbolTable Symbol table to add definitions to
- */
-void ispc::AddBitcodeToModule(const BitcodeLib *lib, llvm::Module *module, SymbolTable *symbolTable) {
-    llvm::Module *bcModule = lib->getLLVMModule();
-
-    // FIXME: this feels like a bad idea, but the issue is that when we
-    // set the llvm::Module's target triple in the ispc Module::Module
-    // constructor, we start by calling llvm::sys::getHostTriple() (and
-    // then change the arch if needed). Somehow that ends up giving us
-    // strings like 'x86_64-apple-darwin11.0.0', while the stuff we
-    // compile to bitcode with clang has module triples like
-    // 'i386-apple-macosx10.7.0'. And then LLVM issues a warning about
-    // linking together modules with incompatible target triples..
-    llvm::Triple mTriple(m->module->getTargetTriple());
-    llvm::Triple bcTriple(bcModule->getTargetTriple());
-    Debug(SourcePos(), "module triple: %s\nbitcode triple: %s\n", mTriple.str().c_str(), bcTriple.str().c_str());
-
-    bcModule->setTargetTriple(mTriple.str());
-    bcModule->setDataLayout(module->getDataLayout());
-
-    if (g->target->isXeTarget()) {
-        // Maybe we will use it for other targets in future,
-        // but now it is needed only by Xe. We need
-        // to update attributes because Xe intrinsics are
-        // separated from the others and it is not done by default
-        lUpdateIntrinsicsAttributes(bcModule);
+llvm::Constant *lFuncAsConstInt8Ptr(llvm::Module &M, const char *name) {
+    llvm::LLVMContext &Context = M.getContext();
+    llvm::Function *F = M.getFunction(name);
+    if (F) {
+        return llvm::ConstantExpr::getBitCast(F, llvm::Type::getInt8PtrTy(Context));
     }
+    return nullptr;
+}
+
+void lRemoveUnused(llvm::Module *M) {
+    llvm::FunctionAnalysisManager FAM;
+    llvm::ModuleAnalysisManager MAM;
+    llvm::ModulePassManager PM;
+    llvm::PassBuilder PB;
+    PB.registerModuleAnalyses(MAM);
+    PM.addPass(llvm::GlobalDCEPass());
+    PM.run(*M, MAM);
+}
 
-    // A hack to move over declaration, which have no definition.
-    // New linker is kind of smart and think it knows better what to do, so
-    // it removes unused declarations without definitions.
-    // This trick should be legal, as both modules use the same LLVMContext.
-    for (llvm::Function &f : *bcModule) {
-        if (f.isDeclaration()) {
-            // Declarations with uses will be moved by Linker.
-            if (f.getNumUses() > 0)
-                continue;
-            module->getOrInsertFunction(f.getName(), f.getFunctionType(), f.getAttributes());
+void lRemoveUnusedPersistentFuntions(llvm::Module *M) {
+    // TODO! make it clean. The idea here is to preserve all persistent
```

Comment:
I'm agree with TODO here :). Looks like this function can be decomposed to several:

- extractUsedFunctions: Extracts the functions that are currently used based on llvm.compiler.used.
- findUsedPersistentGroups: Determines which persistent groups are in use.
- collectPreservedFunctions: Collects the functions that should be preserved based on the used persistent groups.
- updateLLVMUsedGlobal: Updates the llvm.compiler.used global variable with the new list of preserved functions.

Reply:
Done